flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java

```diff
@@ -297,9 +318,279 @@ private static Object handleLogicalTypeSchema(Schema fieldSchema, Object value)
+            mapping.put(
+                    LogicalTypes.localTimestampMillis().getName(),
+                    value -> convertDateTime(value, false));
+            mapping.put(
+                    LogicalTypes.localTimestampMicros().getName(),
+                    value -> convertDateTime(value, true));
+            mapping.put("geography_wkt", AvroSchemaHandler::convertGeography);
+            mapping.put("Json", AvroSchemaHandler::convertJson);
             return mapping.get(logicalTypeString);
         }
+
+        // ---- Utilities to enable Conversions for Logical Types ------------
+        @VisibleForTesting
+        static String convertUUID(Object value) {
+            if (value instanceof UUID) {
+                return ((UUID) value).toString();
+            } else {
+                Preconditions.checkArgument(
+                        value instanceof String, "Expecting a value as String type.");
+                UUID.fromString((String) value);
+                return (String) value;
+            }
+        }
+
+        private static long validateTimestamp(long timestamp) {
+            Timestamp minTs = Timestamp.parseTimestamp("0001-01-01T00:00:00.000000+00:00");
+            Timestamp maxTs = Timestamp.parseTimestamp("9999-12-31T23:59:59.999999+00:00");
+            try {
+                // Validate timestamp from microseconds since EPOCH.
+                Timestamp.ofTimeMicroseconds(timestamp);
+                return timestamp;
+            } catch (IllegalArgumentException e) {
+                throw new IllegalArgumentException(
+                        String.format(
+                                "Invalid Timestamp '%s' Provided."
+                                        + "\nShould be a long value indicating microseconds since Epoch (1970-01-01 00:00:00.000000+00:00) "
+                                        + "between %s and %s",
+                                timestamp, minTs, maxTs));
+            }
+        }
+
+        // BigQuery inputs timestamp as microseconds since EPOCH,
+        // So if we have TIMESTAMP in micros - we convert as it is.
+        // If the TIMESTAMP is in millis - we convert to Micros and then add.
+        @VisibleForTesting
+        static Long convertTimestamp(Object value, boolean micros, String type) {
+            long timestamp;
+            if (value instanceof ReadableInstant) {
+                timestamp = TimeUnit.MILLISECONDS.toMicros(((ReadableInstant) value).getMillis());
+            } else {
+                Preconditions.checkArgument(
+                        value instanceof Long,
+                        String.format(
+                                "Expecting a value as Long type %s. Instead %s was obtained",
+                                type, value.getClass()));
+                timestamp = (micros ? (Long) value : TimeUnit.MILLISECONDS.toMicros((Long) value));
+            }
+            return validateTimestamp(timestamp);
```

Comment:
A method named "validateTimestamp" should not be returning Long. Can we rename this method appropriately? Or let "validateTimestamp" throw an exception and return void. If validate returns successfully then this "convertTimestamp" method can return the timestamp variable.

Reply:
ACK, Fixed.